neqo-common/src/datagram.rs

```diff
@@ -44,17 +41,85 @@ impl Datagram {
     pub fn set_tos(&mut self, tos: IpTos) {
         self.tos = tos;
     }
+
+    #[must_use]
+    pub const fn segment_size(&self) -> usize {
+        self.segment_size
+    }
 }
 
-impl Deref for Datagram {
-    type Target = Vec<u8>;
+impl<D: AsRef<[u8]>> Datagram<D> {
+    pub fn new(
+        src: SocketAddr,
+        dst: SocketAddr,
+        tos: IpTos,
+        d: D,
+        segment_size: Option<usize>,
+    ) -> Self {
+        Self {
+            src,
+            dst,
+            tos,
+            segment_size: segment_size.unwrap_or_else(|| d.as_ref().len()),
```

Comment:
```suggestion
            segment_size: segment_size.unwrap_or(usize::MAX),
```

You always iterate and truncate the last chunk, so this is cheaper.